# Q-A Evaluating metrics

Question Answering Language Models (LLMs) are typically evaluated using a variety of metrics to assess their accuracy and effectiveness in answering questions. Here are some of the common metrics used for evaluating QA LLMs:



**Exact Match (EM)**

The Exact Match (EM) metric is a binary metric used to evaluate the performance of a question-answering system. It measures whether the generated answer by the LLM exactly matches the correct answer. If the generated answer is identical to the correct answer, then the EM score is 1, indicating a perfect match between the generated answer and the correct answer. Otherwise, if the generated answer is different from the correct answer, the EM score is 0, indicating that the LLM did not produce the correct answer.

In [1]:
import re

def normalize_answer(text):
    # Lowercase and remove non-alphanumeric characters
    text = re.sub(r"\W", " ", text.lower())
    # Remove leading and trailing whitespace
    text = text.strip()
    # Collapse multiple whitespace into a single whitespace
    text = re.sub(r"\s+", " ", text)
    return text

def exact_match_score(prediction, ground_truth):
    return int(normalize_answer(prediction) == normalize_answer(ground_truth))

# Example usage
prediction = "I am going to school"
ground_truth = "I am going to college"
em_score = exact_match_score(prediction, ground_truth)
print(f"Exact Match score: {em_score}")

Exact Match score: 0


The EM metric is a simple and straightforward way to evaluate the performance of a question-answering system, and it is often used in conjunction with other metrics, such as the F1 score, to provide a more comprehensive evaluation of the system's performance. However, the EM metric has its limitations, as it is very strict and unforgiving, and it does not take into account partial matches or variations in the correct answer.
For example, if the correct answer is "I am going to college", and the generated answer is "I am going to school", the EM score would be 0, even though the generated answer is very close to the correct answer. Therefore, the EM metric should be used in conjunction with other metrics to provide a more comprehensive evaluation of the performance of a question-answering system.

**F1-Score**

The F1 score is a commonly used metric for evaluating the performance of question-answering systems, particularly in cases where the answer is expected to be a span of text within a larger document. The F1 score is a harmonic mean of precision and recall, which are two other commonly used metrics in natural language processing.The F1 score is the harmonic mean of precision and recall, and it provides a single metric that balances both precision and recall. The F1 score ranges from 0 to 1, with a higher score indicating better performance. An F1 score of 1 means that the LLM generated the correct answer with 100% accuracy and completeness, while a score of 0 means that the LLM did not generate any correct answers.



In [2]:
from sklearn.metrics import f1_score

def f1_score_metric(prediction, ground_truth):
    labels = ['I am going to school', 'I am going to college']
    prediction_label = labels.index(prediction)
    ground_truth_label = labels.index(ground_truth)
    return f1_score([ground_truth_label], [prediction_label], average='weighted')

# Example usage
prediction = "I am going to school"
ground_truth = "I am going to college"
f1 = f1_score_metric(prediction, ground_truth)
print(f"F1 score: {f1}")

F1 score: 0.0


In question answering, the F1 score is often used in conjunction with other metrics, such as the Exact Match (EM) score, to provide a more comprehensive evaluation of the system's performance. The F1 score is particularly useful when the correct answer is expected to be a span of text within a larger document, as it takes into account the position and length of the generated answer in relation to the correct answer.

**BLEU (Bilingual Evaluation Understudy) Score**

BLEU (Bilingual Evaluation Understudy) Score is a metric used to evaluate the quality of machine-generated text against human-generated text. It is commonly used to evaluate the performance of machine translation systems, but it can also be used to evaluate other text generation tasks such as summarization, question answering, and text completion.

The BLEU score is calculated by first computing the precision of the machine-generated text for each n-gram size from 1 to N, where N is typically 4. Precision is the number of n-grams in the machine-generated text that also appear in the human-generated text, divided by the total number of n-grams in the machine-generated text. The precision values are then combined using a geometric mean to produce the final BLEU score.

One limitation of the BLEU score is that it does not take into account the fluency or coherence of the machine-generated text, only its similarity to the human-generated text in terms of n-gram overlap. For this reason, the BLEU score should be used in conjunction with other metrics such as perplexity and human evaluation to provide a more comprehensive evaluation of the quality of the machine-generated text.

In [3]:
import nltk

def bleu_score(prediction, ground_truth):
    reference = [ground_truth.split()]
    candidate = prediction.split()
    return nltk.translate.bleu_score.sentence_bleu(reference, candidate)

# Example usage
prediction = "I am going to school"
ground_truth = "I am going to college"
bleu = bleu_score(prediction, ground_truth)
print(f"BLEU score: {bleu}")

BLEU score: 0.668740304976422


The score ranges from 0 to 1, with higher scores indicating better performance. However, it is important to note that the BLEU score has its limitations, and should be used in combination with other metrics and human evaluation to fully assess the quality and usefulness of machine-generated text. When interpreting BLEU scores, extremely low scores (e.g., on the order of 1e-154 or 1e-231) suggest that the machine-generated text is very dissimilar to the reference text, while scores closer to 1 suggest that the machine-generated text is more similar to the reference text.

**ROUGE (Recall-Oriented Understudy for Gisting Evaluation) Score**

ROUGE (Recall-Oriented Understudy for Gisting Evaluation) Score is a metric used to evaluate the quality of machine-generated text against human-generated text. It is commonly used to evaluate the performance of text summarization systems, but it can also be used to evaluate other text generation tasks such as machine translation and question answering.

ROUGE score measures the overlap between the machine-generated summary and the human-generated summary using n-gram co-occurrence statistics. It is based on the recall of n-gram overlapping between machine-generated text and the reference (human-generated text). The score ranges from 0 to 1, with 1 indicating a perfect match between the two texts. The higher the ROUGE score, the better the machine-generated text is considered to be.



In [4]:
from rouge import Rouge

def rouge_score(prediction, ground_truth):
    rouge = Rouge()
    scores = rouge.get_scores(prediction, ground_truth)
    return scores[0]["rouge-1"]["f"]

# Example usage
prediction = "I am going to school"
ground_truth = "I am going to college"
rouge = rouge_score(prediction, ground_truth)
print(f"ROUGE score: {rouge}")

ROUGE score: 0.7999999950000002


In above case, the ROUGE score is 0.7999999950000002, which is a relatively high score and suggests that there is a good amount of overlap between the model-generated text and reference text. One limitation of the ROUGE score is that it only evaluates the overlap between the machine-generated text and the reference, without considering other important factors such as fluency, coherence, and content preservation. As a result, the ROUGE score should be used in conjunction with other metrics such as BLEU score, perplexity, and human evaluation to provide a more comprehensive evaluation of the quality of the machine-generated text.


**WER (Word Error Rate)**

WER (Word Error Rate) is a metric used for evaluating the performance of speech recognition systems or OCR (Optical Character Recognition) systems. It measures the percentage of words in the recognized text that differ from the words in the reference text. WER is calculated by dividing the total number of word errors (insertions, deletions, and substitutions) by the total number of words in the reference text. 

In [5]:
from jiwer import wer

def wer_metric(prediction, ground_truth):
    return wer(ground_truth, prediction)

# Example usage
prediction = "I am going to school"
ground_truth = "I am going to college"
wer = wer_metric(prediction, ground_truth)
print(f"WER score: {wer}")

WER score: 0.2


A WER (Word Error Rate) score of 0.2 means that 20% of the words in the recognized text are different from the words in the reference text. In other words, there is an error rate of 20%. 